### PR TITLE
Linear stability

### PR DESCRIPTION
**Pull request by roxyboy**
Tests passed for added linear stability code

### CODE REVIEW

**Comment by rabernat** [comment]
@roxyboy this is not working on travis-ci. Your pull request did not pass the checks. Please have a look and try to figure out what is wrong.

If you update your branch and push again, it will update this pull request automatically.

File: oceanmodes/test/test_modes.py [hunk 1]
```diff
@@ -5,8 +5,14 @@
 import numpy as np
 import unittest
 import oceanmodes as modes
+reload(modes)
```
Comment (rabernat) [inline oceanmodes/test/test_modes.py]: You don't want to call reload from within the test script. This is also why the tests fail for python3.

File: oceanmodes/baroclinic.py [hunk 2]
```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
...
     v = v[:,j]
+    
+    #########
+    # w = - 1/(Rd^2 * f0^2)
+    #########
+    Rd = (-w)**-0.5 / f0
+
+    return zf, Rd, v
+
+def instability_analysis_from_N2_profile(z, N2, f0, beta, Nx, Ny, dx, dy, ubar, vbar, etax, etay, sort='LI', num=4, depth=None, **kwargs):
```
Comment (rabernat) [inline oceanmodes/baroclinic.py]: This line is too long. According to the [python style guide](https://www.python.org/dev/peps/pep-0008/), lines should not be longer than 80 characters.

File: oceanmodes/baroclinic.py [hunk 3]
```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
...
+    """Calculate baroclinic neutral modes from a profile of 
+        buoyancy frequency truncated at the bottom.
+    """
+    nz_orig = len(z)
+    z, N2 = _maybe_truncate_above_topography(z, N2)
+    return _instability_analysis_from_N2_profile_raw(
+                                    z, N2, f0, beta, Nx, Ny, dx, dy, ubar, vbar, etax, etay, sort, num, depth=depth, **kwargs)
```
Comment (rabernat) [inline oceanmodes/baroclinic.py:187]: same

File: oceanmodes/baroclinic.py [hunk 4]
```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
...
+
+def _instability_analysis_from_N2_profile_raw(z, N2, f0, beta, Nx, Ny, dx, dy, ubar, vbar, etax, etay, sort='LI', num=4, depth=None, **kwargs):
```
Comment (rabernat) [inline oceanmodes/baroclinic.py]: same

File: oceanmodes/baroclinic.py [hunk 5]
```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
...
+    else:
+        if depth <= z[-1]:
+            raise ValueError('depth should not be less than maximum z')
+
+    dztop = zc[0]
+    dzbot = depth - zc[-1]
+
+    # put the phi points right between the N2 points
+    zf = np.hstack([0, 0.5*(zc[1:]+zc[:-1]), depth ])
+    dzf = np.diff(zf)
```
Comment (rabernat) [inline oceanmodes/baroclinic.py:280]: A lot of this code seems just copied directly from neutral_neutral_modes_from_N2_profile. That means it should be moved into a function of its own.

File: oceanmodes/baroclinic.py [hunk 6]
```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
...
+            and Psi is the streamfunction of the background velocity.
+        eta : array_like
+            The topographic change at the top and bottom.
+        B : array_like
+            The background buoyancy
+        depth : float, optional
+            Total water column depth. If missing will be inferred from z.
+        kwargs : optional
+            Additional parameters to pass to scipy.sparse.linalg.eigs for
+            eigenvalue computation
```
Comment (rabernat) [inline oceanmodes/baroclinic.py:235]: There are lots of function arguments that are not documented in the docstring (nx, ny, dx, dy, etc.)

File: oceanmodes/test/test_modes.py [hunk 7]
```diff
@@ -100,3 +106,152 @@ def test_N2_const_equal_spacing(self):
...
+class TestLinearInstability(unittest.TestCase):
+    """class to test Linear Instability Modes
+    """
+    
+    def test_topography_truncation(self):
+        nz = 10
+        z = np.arange(nz)
+        f = np.ones(nz)
+
+        zt, ft = modes.baroclinic._maybe_truncate_above_topography(z,f)
```
Comment (rabernat) [inline oceanmodes/test/test_modes.py]: Why are you testing these functions again? They are already tested.

File: oceanmodes/test/test_modes.py [hunk 8]
```diff
@@ -100,3 +106,152 @@ def test_N2_const_equal_spacing(self):
...
+        self.assertEqual(len(zt), nz-1)
+
+        f = np.ma.masked_array(f, z==(nz-1))
+        zt, ft = modes.baroclinic._maybe_truncate_above_topography(z,f)
+        self.assertEqual(len(ft), nz-1)
+        self.assertEqual(len(zt), nz-1)
+
+        f.mask[0] = True
+        with self.assertRaises(ValueError):
+            zt, ft = modes.baroclinic._maybe_truncate_above_topography(z,f)
```
Comment (rabernat) [inline oceanmodes/test/test_modes.py]: same

File: oceanmodes/test/test_modes.py [hunk 9]
```diff
@@ -100,3 +106,152 @@ def test_N2_const_equal_spacing(self):
...
+            msg='modes array must be in the right shape')
+
+        self.assertTrue(np.all( np.diff( growth_rate.reshape((growth_rate.shape[0], len(k)*len(l))).imag.max(axis=1) ) < 0.),
+        #self.assertTrue(np.all( max_growth_rate == 0.),
+            msg='imaginary part of modes should be descending')
+
+        #nmodes = len(def_radius)
+        #zero_crossings = np.abs(np.diff(np.sign(bc_modes), axis=0)/2).sum(axis=0)
+        #self.assertListEqual(list(range(nmodes)), list(zero_crossings),
+        #    msg='modes should have the correct number of zero crossings')
```
Comment (rabernat) [inline oceanmodes/test/test_modes.py:221]: Can you think of any tests that will actually verify whether the output is correct?

**Comment by rabernat** [comment]
@roxyboy I made lots of line comments on your pull request.

File: oceanmodes/test/test_modes.py (hunk 9 above)
Comment (roxyboy) [inline oceanmodes/test/test_modes.py:221]: Hi Ryan,

Thank you very much for your feedbacks.
I can’t seem to come up with a test to see whether the outputs are correct.
Could I discuss this with you tomorrow?

Best,
Takaya

Comment (rabernat) [inline oceanmodes/test/test_modes.py:221]: What about the Eady problem? Charney problem?

**Comment by rabernat** [comment]
@roxyboy for now, why don't you just address the other issues I found (line length, python 3 compatibility, etc.). We can leave the other tests for a future pull request.

**Pull request merged**